src/rgw/rgw_flight.h

```diff
@@ -0,0 +1,222 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#pragma once
+
+#include <map>
+#include <mutex>
+#include <atomic>
+
+#include "include/common_fwd.h"
+#include "common/ceph_context.h"
+#include "common/Thread.h"
+#include "common/ceph_time.h"
+#include "rgw_frontend.h"
+#include "arrow/type.h"
+#include "arrow/flight/server.h"
+#include "arrow/util/string_view.h"
+
+#include "rgw_flight_frontend.h"
+
+
+#define INFO_F(dp)   ldpp_dout(&dp, 20) << "INFO: " << __func__ << ": "
+#define STATUS_F(dp) ldpp_dout(&dp, 10) << "STATUS: " << __func__ << ": "
+#define WARN_F(dp)   ldpp_dout(&dp,  0) << "WARNING: " << __func__ << ": "
+#define ERROR_F(dp)  ldpp_dout(&dp,  0) << "ERROR: " << __func__ << ": "
+
+#define INFO   INFO_F(dp)
+#define STATUS STATUS_F(dp)
+#define WARN   WARN_F(dp)
+#define ERROR  ERROR_F(dp)
+
+
+namespace arw = arrow;
+namespace flt = arrow::flight;
+
+
+struct req_state;
+
+namespace rgw::flight {
+
+  static const coarse_real_clock::duration lifespan = std::chrono::hours(1);
+
+  struct FlightData {
+    FlightKey key;
+    // coarse_real_clock::time_point expires;
+    std::string uri;
+    std::string tenant_name;
+    std::string bucket_name;
+    rgw_obj_key object_key;
+    // NB: what about object's namespace and instance?
+    uint64_t num_records;
+    uint64_t obj_size;
+    std::shared_ptr<arw::Schema> schema;
+    std::shared_ptr<const arw::KeyValueMetadata> kv_metadata;
+
+    rgw_user user_id; // TODO: this should be removed when we do
+		      // proper flight authentication
+
+    FlightData(const std::string& _uri,
+	       const std::string& _tenant_name,
+	       const std::string& _bucket_name,
+	       const rgw_obj_key& _object_key,
+	       uint64_t _num_records,
+	       uint64_t _obj_size,
+	       std::shared_ptr<arw::Schema>& _schema,
+	       std::shared_ptr<const arw::KeyValueMetadata>& _kv_metadata,
+	       rgw_user _user_id);
+    FlightData(const FlightData&) = default;
+    FlightData(FlightData&&) = default;
+    FlightData& operator=(const FlightData&) = default;
+    FlightData& operator=(FlightData&&) = default;
```

Comment:
these FlightData ops would be defaulted already. see https://en.cppreference.com/w/cpp/language/rule_of_three#Rule_of_zero